memory/src/memory.cpp

```diff
@@ -0,0 +1,221 @@
+        // Read next PTE
+        PhysAddr pte_pa = table_ppn * PAGE_SIZE + getVPN(va, i) * sizeof(PTE);
+
+        if (m_phys_memory.read(pte_pa, pte).status !=
+            PhysMemory::AccessStatus::OK) {
+            return ACCESS_FAULT_RES;
+        }
+
+        flags = pte;
+
+        // Check PTE validity
+        if (flags.v() || (!flags.r() && flags.w()) ||
+            (pte & PTE_RESERWED_MASK)) {
+            return PAGE_FAULT_RES;
+        }
+
+        // Leaf PTE found
+        if (flags.r() || flags.x()) {
+            break;
+        }
+
+        // Leaf PTE not found
+        if (i == 0) {
+            return PAGE_FAULT_RES;
+        }
+
+        table_ppn = bit::getBitField(PTE_PPN_HI, PTE_PPN_LO, pte);
+    }
+
+    // Check if the access is allowed
+    if (!checkLeafFlags(priv_level, access_type, flags, m_mstatus64.getMXR(),
+                        m_mstatus64.getSUM())) {
+        return PAGE_FAULT_RES;
+    }
+
+    // Check superpage alignment
+    if (i > 0) {
+        static constexpr bit::BitIdx lo = 10;
+
+        if (bit::getBitField<PTE>(pte, PPN_BIT_STEP * i - 1 + lo, lo)) {
+            return PAGE_FAULT_RES;
+        }
+    }
+
+    // A and D bits are set to 1 by default
+    SIM_ASSERT(flags.d() && flags.a());
+
+    // The translation is successful
+    return {Status::OK, calcPhysAddr(pte, va, i)};
+}
+
+NODISCARD SimpleMemoryMapper::MapStatus
+SimpleMemoryMapper::map(MemoryMapping mapping) noexcept {
+    // Mappings for table region pages are forbidden
+    if (mapping.ppn() >= m_table_region_begin &&
+        mapping.ppn() < m_table_region_end) {
+        return MapStatus::MAPPING_WITHIN_TABLE_REGION;
+    }
+
+    VirtAddr va = mapping.vpn() * PAGE_SIZE;
+
+    size_t i = modeToLevels(m_mode) - 1;
+
+    // Write mapping PTEs
+    for (PhysAddr table_ppn = m_table_region_begin;; --i) {
+        // Read next PTE
+        PhysAddr pte_pa = table_ppn * PAGE_SIZE + getVPN(va, i) * sizeof(PTE);
+
+        PTE pte = 0;
+        if (m_phys_memory.read(pte_pa, pte).status !=
+            PhysMemory::AccessStatus::OK) {
+            return MapStatus::PHYS_MEMORY_ERROR;
+        }
+
+        // Read PTE is not valid => add new PTE
+        if (!PTEFlags(pte).v()) {
+            if (i != 0) {
+                // Add new page table
+                if (m_curr_table == m_table_region_end) {
+                    return MapStatus::TABLE_REGION_END;
+                }
+
+                PPN new_table = m_curr_table++;
+                pte = createPTE(new_table, PTEFlags::FlagsMask::V_MASK);
+            } else {
+                // Add leaf PTE
+                pte = createPTE(mapping.ppn(), mapping.flags());
+            }
+
+            // Write PTE
+            if (m_phys_memory.write(pte_pa, pte).status !=
+                PhysMemory::AccessStatus::OK) {
+                return MapStatus::PHYS_MEMORY_ERROR;
+            }
+
+            // Mapping created
+            if (i == 0) {
```

Comment:
Maybe put this check at the beginning of func?

Reply:
Page table walk with PTEs insertions is required prior this check